### PR TITLE
Implement SA1314 (TypeParameterNamesMustBeginWithT)

### LINKED ISSUE #739: Implement SA1314: TypeParameterNamesMustBeginWithT
I'd like to propose new naming rule for generic parameters. Name should start with "T", second character (if it exists) should be upper case letter or digit. AFAIK Whole BCL follows this convention. 

Few valid names: "T", "T1", "TKey", "TValue" 

Few invalid names: "t", "Tkey", "Key" 

This rule was originally proposed by Andy Reeves here:
http://stylecop.codeplex.com/workitem/6891

### PR DESCRIPTION
**Pull request by nbarbettini**
Fixes #739. Includes code fix, tests, and documentation.

Let me know if there's any test scenarios I didn't think of!

### CODE REVIEW

**Comment by sharwell** [comment]
:exclamation: naming rules have IDs with the form SA13xx.

**Comment by vweijsters** [comment]
Type parameters would be preferred for me

**Comment by nbarbettini** [comment]
I wasn't sure myself; I'd be totally fine with changing it to "type parameters" across the board.

**Comment by nbarbettini** [comment]
Fixed both the ID and the rule name. :+1:

File: StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/Helpers/RenameHelper.cs [hunk 1]
```diff
@@ -52,6 +52,20 @@ internal static class RenameHelper
 
             var containingSymbol = symbol.ContainingSymbol;
 
+            if (symbol.Kind == SymbolKind.TypeParameter)
```
Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/Helpers/RenameHelper.cs]: We cant rename the type parameter if there is already a type with the same name

File: StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/NamingRules/SA1314CodeFixProvider.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                    diagnostic);
+            }
+
+            return SpecializedTasks.CompletedTask;
+        }
+
+        private static async Task<Solution> CreateChangedSolutionAsync(Document document, Diagnostic diagnostic, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(diagnostic.Location.SourceSpan.Start);
```
Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/NamingRules/SA1314CodeFixProvider.cs]: Do we also want to make the first letter after T capital?

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/NamingRules/SA1314UnitTests.cs [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+        var key = typeof(TKey);
+    }
+}
+";
+
+            await this.VerifyCSharpFixAsync(testCode, fixedCode).ConfigureAwait(false);
+        }
+
+        [Fact]
+        public async Task TestGenericParameterStartsWithLowerTAsync()
```
Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/NamingRules/SA1314UnitTests.cs]: Do we want to be smarter in this case?

**Comment by pdelvo** [comment]
Maybe we should include a test to verify that this also works for type parameters in method signatures

File: StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/Helpers/RenameHelper.cs (hunk 1 above)
Comment (nbarbettini) [inline StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/Helpers/RenameHelper.cs]: You mean (for example) if there was already a class named `TKey`?

File: StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/NamingRules/SA1314CodeFixProvider.cs (hunk 2 above)
Comment (nbarbettini) [inline StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/NamingRules/SA1314CodeFixProvider.cs]: I followed the example of SA1302, which simply adds an I and doesn't touch the rest of the token. This makes sense, though; I could implement this change if others agree.

File: StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/Helpers/RenameHelper.cs (hunk 1 above)
Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/Helpers/RenameHelper.cs]: Exactly

File: StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/NamingRules/SA1314CodeFixProvider.cs (hunk 2 above)
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/NamingRules/SA1314CodeFixProvider.cs]: I don't mind just adding the `T` prefix.

**Comment by nbarbettini** [comment]
@pdelvo Added a test case for type parameters in method signatures, and also refactored the logic in `RenameHelper` to handle conflicts with other types.

**Pull request merged**